### ISSUE TITLE
Changing "bind-address" of kube-api

### ISSUE THREAD

**Issue by gakio12** [k0sproject/k0s#957]
I have systems with multiple network interfaces and I would like to initialize the cluster on only one of the interfaces, say 192.168.1.1.

## TLDR
Adding `spec.api.extraArgs.bind-address = 192.168.1.1` (and adding to sans) in k0s config successfully causes the kube-api server to only bind on 192.168.1.1. However, when changing the `/var/lib/k0s/pki/admin.conf`s `clusters[0].cluster.server = https://192.168.1.1:6443`, `k0s token create --role=worker` results in a `TLS handshake timeout`. 

## The rest
This issue is more a question of how to configure k0s to achieve what I am looking to do. I don't want the kube-api server listening on every interface in this case, but changing the bind-address breaks being able to access the cluster. I know it is a certificate issue, I'm just not sure where.

Here is the relevant section in my k0s.yaml file:
```yaml
spec:
  api:
    address: 192.168.1.1
    port: 6443
    k0sApiPort: 9443
    externalAddress: 192.168.1.1
    extraArgs:
      bind-address: 192.168.1.1
    sans:
    - 192.168.1.1
```

For context, this is a single node control plane cluster that manages airgapped nodes that are behind a specific network interface.

**Comment by jnummelin** [comment]
> TLS handshake timeout

Sounds more like an issue of the API not really getting up rather than "wrong" cert info. So check that you can hit the API with curl, e.g. `curl -k -v https://192.168.1.1:6443`. Does `k0s kubectl get node` give any different error?

> However, when changing the /var/lib/k0s/pki/admin.confs clusters[0].cluster.server = https://192.168.1.1:6443

hmm, that _should_ do the trick for the k0s token commands. BUT just a side note that there's bunch of other component access configs which has `localhost` as the address.

The above being said, I think we really need to make the `bind-address` a "first class citizen" in the config so we can really pass it along to all needed components too.

**Comment by gakio12** [comment]
> Sounds more like an issue of the API not really getting up
It looks like the API is coming up, curl does work, but only until the API is eventually stopped after the health-check continually fails. So in this case, the health-check address is still set to `localhost`, would need to change this as well before this works. I tried the flag `--api-server=192.168.1.1` for `k0s controller`, but that didn't change the health check address.

**Comment by jnummelin** [comment]
> I tried the flag --api-server=192.168.1.1 for k0s controller, but that didn't change the health check address.

yeah, many things internal in k0s is hardwired to use localhost. :(

Thus making the `bindAddress` as a k0s.yaml configurable options is the proper fix for this as that allows us to pass that on to all needed places.

**Comment by gakio12** [comment]
I'm working on adding the `bindAddress` feature, for the most part it is in place. Adding documentation on it, I'm not sure if I am duplicating what `address` _should_ have done, or if the documentation for `address` was misleading in the first place. `address` is defined as "Local address on which to bind an API...", but it never really honored this by passing it to `--bind-address` to the API server, causing it to bind on `0.0.0.0`.

@jnummelin Do you think `bindAddress` should be added and change the documentation on `address`, or should `address` be passed to the API server `--bind-address` flag?

**Comment by gakio12** [comment]
Can I get this re-opened? I am planning on finishing this up soon now that my schedule has opened back up.

**Comment by mtb-xt** [comment]
Hey everyone, I wonder if this will ever be fixed? I can't get kube-apiserver to _not_ listen on 0.0.0.0 with extraArgs and it's not working for me :/
```
apiVersion: k0s.k0sproject.io/v1beta1
kind: ClusterConfig
metadata:
  creationTimestamp: null
  name: k0s
spec:
  api:
    address: 172.21.34.6
    extraArgs: 
      bind-address: 172.21.34.6
    k0sApiPort: 9443
    port: 6443
    sans:
    - 172.21.34.6
    - 172.21.34.230
```
and still getting 
```
Sep 09 12:00:15 sentinel k0s[15744]: time="2022-09-09 12:00:15" level=info msg="E0909 12:00:15.797354   15934 run.go:74] \"command failed\" err=\"failed to create listener: failed to listen on 0.0.0.0:6443: listen tcp 0.0.0.0:6443: bind: address already in use\"" component=kube-apiserver
```
because something is still passing 
`FLAG: --bind-address=\"0.0.0.0\"" component=kube-apiserver`

**Comment by jnummelin** [comment]
hey, we're still planning to fix this. But properly supporting bind-addresses means we should support it in all components that run on controller nodes. While most components have the options available, konnectivity-server unfortunately did not have. So I did an upstream PR to get those available: https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/363

Unfortunately this work is still "blocked" as konnectivity folks have not tagged a release containing the needed fixes.

The WIP PR is k0s side is here: https://github.com/k0sproject/k0s/pull/1880

**Comment by onedr0p** [comment]
I believe I am seeing this issue because the API server is trying to bind to a non-valid interface via the default `0.0.0.0`:

```
6: idrac: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN group default qlen 1000
    link/ether cc:48:3a:08:fa:72 brd ff:ff:ff:ff:ff:ff
    inet 169.254.1.2/16 brd 169.254.255.255 scope global idrac
       valid_lft forever preferred_lft forever
    inet6 fde1:53ba:e9a0:de11:ce48:3aff:fe08:fa72/64 scope global dynamic mngtmpaddr 
       valid_lft 86296sec preferred_lft 14296sec
    inet6 fe80::ce48:3aff:fe08:fa72/64 scope link 
       valid_lft forever preferred_lft forever
```

The error I am seeing in the log:

```
Feb 02 11:03:56 expanse k0s[301425]: time="2024-02-02 11:03:56" level=info msg="E0202 11:03:56.332223  301735 controller.go:159] unable to sync kubernetes service: Endpoints \"kubernetes\" is invalid: subsets[0].addresses[0].ip: Invalid value: \"169.254.1.2\": may not be in the link-local range (169.254.0.0/16, fe80::/10)" component=kube-apiserver stream=stderr
```
